aws/hll/aws-s3-transfer-manager/src/download/header.rs

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use core::fmt;
+use std::str::FromStr;
+
+use crate::error;
+
+/// Representation of `Range` header.
+/// NOTE: S3 only supports a single bytes range this is a simplified representation
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct Range(pub(crate) ByteRange);
+
+impl Range {
+    /// Create a range from the given byte range
+    pub(crate) fn bytes(rng: ByteRange) -> Self {
+        Self(rng)
+    }
+
+    /// Create a range from the inclusive start and end offsets
+    pub(crate) fn bytes_inclusive(start: u64, end: u64) -> Self {
+        Range::bytes(ByteRange::Inclusive(start, end))
+    }
+}
+
+impl fmt::Display for Range {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "bytes={}", self.0)
+    }
+}
+
+impl From<Range> for String {
+    fn from(value: Range) -> Self {
+        format!("{}", value)
+    }
+}
+
+impl FromStr for Range {
+    type Err = error::TransferError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut iter = s.splitn(2, '=');
+        match (iter.next(), iter.next()) {
+            (Some("bytes"), Some(range)) => {
+                if range.contains(',') {
+                    // TODO - error S3 doesn't support multiple byte ranges
+                    Err(error::invalid_meta_request(format!(
+                        "multiple byte ranges not supported for range header {}",
+                        s
+                    )))
+                } else {
+                    let spec = ByteRange::from_str(range).map_err(|_| {
+                        error::invalid_meta_request(format!("invalid range header {}", s))
+                    })?;
+                    Ok(Range(spec))
+                }
+            }
+            _ => Err(error::invalid_meta_request(format!(
+                "unsupported byte range header format {}",
```

Comment:
If an error would make its way to a user, I think they should include calls to action that say what the user should do about their mistake.

Reply:
Yeah, maybe including an URL to the RFC (or just RFC number and the section name) for supported range formats could be helpful.